pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerMemoryLeakTest.java

```diff
@@ -0,0 +1,376 @@
+        final String topicName = BrokerTestUtil.newUniqueName("persistent://public/default/tp_");
+        admin.topics().createNonPartitionedTopic(topicName);
+        ProducerImpl<String> producer = (ProducerImpl<String>) pulsarClient.newProducer(Schema.STRING).topic(topicName)
+                .intercept(
+
+                new ProducerInterceptor() {
+                    @Override
+                    public void close() {
+                        if (method.equals("close")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                    }
+
+                    @Override
+                    public boolean eligible(Message message) {
+                        if (method.equals("eligible")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                        return false;
+                    }
+
+                    @Override
+                    public Message beforeSend(Producer producer, Message message) {
+                        if (method.equals("beforeSend")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                        return message;
+                    }
+
+                    @Override
+                    public void onSendAcknowledgement(Producer producer, Message message, MessageId msgId,
+                                                      Throwable exception) {
+                        if (method.equals("onSendAcknowledgement")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+
+                    }
+                }).create();
+
+        MsgPayloadTouchableMessageBuilder<String> msgBuilder = newMessage(producer);
+        try {
+            msgBuilder.value("msg-1").sendAsync().get(3, TimeUnit.SECONDS);
+            // It may throw error.
+        } catch (Exception ex) {
+            assertTrue(ex.getMessage().contains("Mocked"));
+        }
+
+        // Verify: message payload has been released.
+        producer.close();
+        assertEquals(msgBuilder.payload.refCnt(), 1);
+
+        // cleanup.
+        msgBuilder.release();
+        assertEquals(msgBuilder.payload.refCnt(), 0);
+        admin.topics().delete(topicName);
+    }
+
+    private <T> MsgPayloadTouchableMessageBuilder<T> newMessage(ProducerImpl<T> producer){
+        return new MsgPayloadTouchableMessageBuilder<T>(producer, producer.schema);
+    }
+
+    private static class MsgPayloadTouchableMessageBuilder<T> extends TypedMessageBuilderImpl {
+
+        public volatile ByteBuf payload;
+
+        public <T> MsgPayloadTouchableMessageBuilder(ProducerBase producer, Schema<T> schema) {
+            super(producer, schema);
+        }
+
+        @Override
+        public Message<T> getMessage() {
+            beforeSend();
+            MessageImpl<T> msg = (MessageImpl<T>) super.getMessage();
```

Comment:
`super.getMessage()` already calls `beforeSend`, so I think you don't need to make any change to `beforeSend`

Reply:
Removed